### ISSUE TITLE
Enabling session expired notification prevents seamless logout button

### ISSUE THREAD

**Issue by mvysny** [vaadin/flow#9647]
Enabling the session-expired notification via `CustomizedSystemMessages.setSessionExpiredNotificationEnabled(true)` prevents me from having a seamless logout button since the "Session Expired" message flashes briefly when logging out.

The logout button is implemented as usual:

```java
        Button button = new Button("Logout",
                e -> {
                    UI.getCurrent().getPage().setLocation("/");
                    VaadinSession.getCurrent().close();
                    VaadinSession.getCurrent().getSession().invalidate();
                } );
```

### Expected behavior

Pressing the button should reload the page without showing the "session expired" notification.

### Actual behavior

Pressing the button will briefly show a red "session-expired" notification then reloads the page.

### Workarounds

1. Using `UI.getCurrent().getPage().open("/", "_self");` instead of `setLocation()` as suggested at https://github.com/vaadin/framework/issues/4255 doesn't help - the "session expired" notification still flashes.
2. Disable the session expired notification.

### Versions:
    - Vaadin / Flow version: 14.4.4 / 2.4.3
    - Java version: 11

**Comment by mvysny** [comment]
Example project which reproduces the issue
[skeleton-starter-flow.zip](https://github.com/vaadin/flow/files/5689021/skeleton-starter-flow.zip)

**Comment by denis-anisimov** [comment]
Hmmmm....
That should have been fixed.

At least we have a test for this case : `com.vaadin.flow.uitest.ui.LogoutView`.
Though if I recall correctly the main issue there was related to PUSH: it refused to redirect at all.
The reason was that session expired message is handled on the client without knowing that the app is closed at that point.
As a fix the application close has been added to the JS executed in `Page::open`: 
`"if ($1 == '_self') this.stopApplication(); window.open($0, $1)"`.

I think now it's a race condition because session is closed before redirect is done: the session expired message is handled on the client side _before_ the redirect JS is executed.
I think this is the case.
And I assume the steps to reproduce should sometimes work and sometimes doesn't.

**Comment by mshabarov** [comment]
This issue has been fixed by making JavaScript commands execution happens before "session expire" and "app error" message handling in `MessageHandler` class:
Before:
```java
            ValueMap meta = valueMap.getValueMap("meta");

            if (meta != null) {
                // Message handling code
            }

            // Invokes all the flush listeners, including JS commands execution 
            Reactive.flush();
```

After:
```java
            // Invokes all the flush listeners, including JS commands execution 
            Reactive.flush();

            ValueMap meta = valueMap.getValueMap("meta");

            if (meta != null) {
                // Message handling code. Execute if UI has been TERMINATED
            }
```
No `LogoutIT` failures (https://github.com/vaadin/flow/issues/9705) were observed then during few days.
But then, this fix was reverted because it revealed the issue with `SessionCloseLogoutIT`: when the redirect is made by execution of the custom JS `window.location.href="url-to-redirect"`, and not by `Page::setLocation`, this test fails and shows that the redirect actually does not work. The application just refreshes the page and that's it.

Then, I made a research on what is the root cause of this and found the following:

1. Redirect doesn't work even if disable the Push by commenting out the annotation `@Push(transport = Transport.LONG_POLLING)`
2. Redirect does work, if the UI is TERMINATED just before the redirect, like this: `this.stopApplication(); window.location.href="url-to-redirect"`.

Point 2 gave me a clue that I could rework the code to implement the following sequence:

```java
            ValueMap meta = valueMap.getValueMap("meta");

            if (meta != null) {
                 // TERMINATE UI if "session expire" or "app error" message
            }
            // Invokes all the flush listeners, including JS commands execution 
            Reactive.flush();

            if (meta != null) {
                // Message handling code.
            }
```

But, unfortunately, it didn't work and the test was still failing.

Then, after adding some trace logging, I found that exactly this block of code in `SystemErrorHandler::handleUnrecoverableError` causes the redirect to be canceled:

```java
        if (caption == null && message == null && details == null) {
            if (!isWebComponentMode()) {
                // This does the page refresh and cancels the redirect request:
                WidgetUtil.redirect(url);
            }
            return;
        }
```

The browser console shows it:

<img width="1790" alt="Screenshot 2021-03-11 at 11 11 45" src="https://user-images.githubusercontent.com/61410877/110763558-b2997100-825a-11eb-8db9-b2601a102c4c.png">

So, at the moment, I have no clues why:
1. Terminating the UI before `Reacive.flush()` (and then handle the message) does not work, however `this.stopApplication()` does work
2. Refresh of the page takes the old page and cancels the redirect request. Probably, it is how the browser works.
3. The locations obtained by calling `Browser.getWindow().getLocation().getHref()` before and after redirect execution are the same. However, I suppose it should be different.

**Comment by mshabarov** [comment]
Reopened, since the fix has been reverted due to regression.

**Comment by denis-anisimov** [comment]
This is a mixture of various issues so I would like to have here a resume of the problem:
- this issue is fixed by  #10390
- there is a similar ticket #9705 which is quite close to this ticket except the LogoutView doesn't use `CustomizedSystemMessages`
- There is another very similar test `SessionCloseLogoutIT` which is initially broken and this fact allows it to pass in the branches where it's enabled (see #10389: it enables the test in the `master` ).

Details are here:
- `LogoutView` from #9705 doesn't use any system messages which means that the client side uses `WidgetUtil.redirect(url)` logic on session expired.
- This ticket uses `CustomizedSystemMessages` which enables another branch of the client side logic : it shows an error message instead of redirect .

That's why the behavior in the #9705 is different and randomly fails.

The `MessageHandler`  (on the client side , `flow-client`) checks first whether there is session expired message and then it flushes all the command via `Reactive.flush()`.

As a result : there is no any reasons to execute "session expired"  logic _before_ other commands but in fact it's better to execute session expired _after_ since there can be commands which should "cancel" (or hide the effect of) session expired execution.

There is a `SessionCloseLogoutIT`  which failed after changes in #10390 (that was the reason why the fix has been reverted).
If fails because of totally different reasons.
The test as initially broken because of https://github.com/vaadin/flow/issues/8177.
The latter ticket is exactly about this usecase: Long Polling Push + redirect.

 `SessionCloseLogoutView` has no  `CustomizedSystemMessages`  which means that `WidgetUtil.redirect(url)` branch is used.
And this execution is cancelled because of the #8177 bug.
Then JS execution is handled properly.
So  `SessionCloseLogoutIT`  works because it's buggy.
When  #10390 is applied the `window.location.href=`  is cancelled instead of `WidgetUtil.redirect(url)` and this is exactly the  #8177.
So `SessionCloseLogoutIT`  should not use `window.location.href=`  but it should use `setLocation` instead (which terminates the UI).
This is fixed here : https://github.com/vaadin/flow/pull/10389
